File: samples/Sounds/src/SoundsSample.cpp
// Oak Nut include
#include <onut/Curve.h>
#include <onut/Font.h>
#include <onut/Input.h>
#include <onut/Log.h>
#include <onut/Music.h>
#include <onut/Renderer.h>
#include <onut/Settings.h>
#include <onut/SpriteBatch.h>
#include <onut/Sound.h>

// STL
#include <cmath>

OSoundInstanceRef pLoopingSound;
OMusicRef pMusic;
OSoundRef pNotes[8];

static const double NOTE_FREQUENCIES[8] = {
    261.63, // C4
    293.66, // D4
    329.63, // E4
    349.23, // F4
    392.00, // G4
    440.00, // A4
    493.88, // B4
    523.25, // C5
};

void initSettings()
{
    oSettings->setGameName("Sounds Sample");
    oSettings->setIsResizableWindow(true);
}

void init()
{
    pMusic = OGetMusic("music.ogg");

    pLoopingSound = OCreateSoundInstance("looping.wav");
    pLoopingSound->setLoop(true);

    // Create sounds from custom data
    float* pSampleBuffer = new float[50000];
    for (int note = 0; note < 8; ++note)
    {
        for (int i = 0; i < 50000; ++i)
        {
            // Create a sine wave of a standard C4 note
            pSampleBuffer[i] = (float)std::sin((double)i / 44100.0 * NOTE_FREQUENCIES[note] * 3.1415926535897932384626433832795 * 2.0);

            // Fade out
            float fadeOut = (float)((double)(50000 - i) / 50000.0);
            fadeOut = std::pow(fadeOut, 4.0f);
            pSampleBuffer[i] *= fadeOut;
        }
        pNotes[note] = OSound::createFromData(pSampleBuffer, 50000, 1, 44100);
    }
    delete[] pSampleBuffer;
}

void update()
{
    float volume = 1.f;
    float balance = 0.f;
    float pitch = 1.0f;

    if (OInputPressed(OKeyLeft)) balance -= 1.f;
    if (OInputPressed(OKeyRight)) balance += 1.f;
    if (OInputPressed(OKeyUp)) pitch = 2.f;
    if (OInputPressed(OKeyDown)) pitch = 0.5f;

    if (OInputJustPressed(OKeyF1))
    {
        OPlaySound("48000_8.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF2))
    {
        OPlaySound("48000_16.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF3))
    {
        OPlaySound("48000_24.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF4))
    {
        OPlaySound("48000_32.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF5))
    {
        OPlaySound("44100_32.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF6))
    {
        OPlaySound("22050_32.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF7))
    {
        OPlaySound("11025_32.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF8))
    {
        OPlaySound("8000_32.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKeyF9))
    {
        OPlaySound("48000_32_2ch.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKey1))
    {
        OPlaySound("sound1.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKey4))
    {
        OPlayRandomSound({"sound1.wav", "sound2.wav", "sound3.wav", "sound4.wav", "sound5.wav"}, 
                         volume, balance, pitch);
    }
    if (OInputJustPressed(OKey5))
    {
        OPlaySound("sound5.wav", volume, balance, pitch);
    }
    if (OInputJustPressed(OKey6))
    {
        if (pLoopingSound->isPlaying())
        {
            pLoopingSound->pause();
        }
        else
        {
            pLoopingSound->setBalance(balance);
            pLoopingSound->setPitch(pitch);
            pLoopingSound->play();
        }
    }
    if (OInputJustPressed(OKey7))
    {
        OPlaySoundCue("soundCue.cue", volume, balance, pitch);
    }
    if (OInputJustPressed(OKey9))
    {
        if (pMusic->isPlaying())
        {
            OLog("Stop Music");
            pMusic->stop();
        }
        else
        {
            OLog("Start Music");
            pMusic->play();
        }
    }
    if (OInputJustPressed(OKeySpaceBar))
    {
        if (!pMusic->isPaused()) pMusic->pause();
        else pMusic->resume();
    }
    if (OInputJustPressed(OKeyQ)) pNotes[0]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyW)) pNotes[1]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyE)) pNotes[2]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyR)) pNotes[3]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyT)) pNotes[4]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyY)) pNotes[5]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyU)) pNotes[6]->play(volume, balance, pitch);
    if (OInputJustPressed(OKeyI)) pNotes[7]->play(volume, balance, pitch);
}

void render()
{
    // Clear to black
    oRenderer->clear({0, 0, 0, 1});

    // Draw info
    auto pFont = OGetFont("font.fnt");

    oSpriteBatch->begin();

    pFont->draw("Press ^990F1^999 to play 48000 hz, 8 bits", { 400, 10});
    pFont->draw("Press ^990F2^999 to play 48000 hz, 16 bits", { 400, 30});
    pFont->draw("Press ^990F3^999 to play 48000 hz, 24 bits", { 400, 50 });
    pFont->draw("Press ^990F4^999 to play 48000 hz, 32 bits", { 400, 70 });
    pFont->draw("Press ^990F5^999 to play 44100 hz, 32 bits", { 400, 90 });
    pFont->draw("Press ^990F6^999 to play 22050 hz, 32 bits", { 400, 110 });
    pFont->draw("Press ^990F7^999 to play 11025 hz, 32 bits", { 400, 130 });
    pFont->draw("Press ^990F8^999 to play 8000 hz, 32 bits", { 400, 150 });
    pFont->draw("Press ^990F9^999 to play 48000 hz, 32 bits, 2 channels", { 400, 170 });

    pFont->draw("Press ^9901^999 to play sound1", {10, 10});
    pFont->draw("Press ^9904^999 to play a random sound", {10, 70});
    pFont->draw("Press ^9905^999 to play sound5", {10, 90});
    if (pLoopingSound->isPlaying())
    {
        pFont->draw("Press ^9906^999 to start/stop looping sound (^099Playing^999)", {10, 110});
    }
    else
    {
        pFont->draw("Press ^9906^999 to start/stop looping sound", {10, 110});
    }
    pFont->draw("Press ^9907^999 to play cue file", {10, 130});
    pFont->draw("Press ^990qwertyui^999 to do music", {10, 150});
    pFont->draw("Press ^9909^999 to play/stop music (^990SpaceBar^999 to pause/resume)", {10, 170});

    pFont->draw("Hold ^990Left Arrow^999 to on left channel", { 10, OScreenHf - 50 });
    pFont->draw("Hold ^990Right Arrow^999 to on right channel", { 10, OScreenHf - 30 });

    pFont->draw("Hold ^990Up Arrow^999 to play at 200% speed", { 10, OScreenHf - 100 });
    pFont->draw("Hold ^990Down Arrow^999 to play at 50% speed", { 10, OScreenHf - 80 });

    oSpriteBatch->end();
}

void postRender()
{
}

File: samplesJS/Sounds/assets/scripts/SoundsSample.js
var font = getFont("font.fnt");
var music = getMusic("music.ogg");
var sound5 = getSound("sound5.wav");
var loopingSound = createSoundInstance("looping.wav");
loopingSound.setLoop(true);

var notes = [];

var NOTE_FREQUENCIES = [
    261.63, // C4
    293.66, // D4
    329.63, // E4
    349.23, // F4
    392.00, // G4
    440.00, // A4
    493.88, // B4
    523.25, // C5
];

// Create sounds from custom data.
// Not recommanded to do in Javascript. This is extremely slow.
var sampleBuffer = new Float32Array(25000);
for (var note = 0; note < 8; ++note)
{
    var noteFreq = NOTE_FREQUENCIES[note];
    var period = 1 / 44100 * Math.PI * 2 * noteFreq;
    for (var i = 0; i < 25000; ++i)
    {
        // Create a sine wave of a standard C4 note
        var val = Math.sin(i * period);

        // Fade out
        var fadeOut = (25000 - i) / 25000;
        fadeOut = Math.pow(fadeOut, 4);

        sampleBuffer[i] = val * fadeOut;
    }
    notes[note] = Sound.createFromData(sampleBuffer, 25000, 1, 44100);
}
sampleBuffer = null; // Free memory

function update() {
    var volume = 1;
    var balance = 0;

    if (Input.isDown(Key.LEFT)) balance -= 1;
    if (Input.isDown(Key.RIGHT)) balance += 1;

    if (Input.isJustDown(Key._1)) {
        playSound("sound1.wav", volume, balance);
    }
    if (Input.isJustDown(Key._2)) {
        playSound("sound1.wav", volume, balance, .5);
    }
    if (Input.isJustDown(Key._3)) {
        playSound("sound1.wav", volume, balance, 2);
    }
    if (Input.isJustDown(Key._4)) {
        playRandomSound(["sound1.wav", "sound2.wav", "sound3.wav", "sound4.wav", "sound5.wav"],
            volume, balance);
    }
    if (Input.isJustDown(Key._5)) {
        // Play using the resource directly
        sound5.play(volume, balance);
    }
    if (Input.isJustDown(Key._6))
    {
        if (loopingSound.isPlaying())
        {
            loopingSound.pause();
        }
        else
        {
            loopingSound.play();
        }
    }
    if (Input.isJustDown(Key._7)) {
        playSoundCue("soundCue.cue", volume, balance);
    }
    if (Input.isJustDown(Key._9)) {
        if (music.isPlaying()) {
            print("Stop Music");
            music.stop();
        }
        else {
            print("Start Music");
            music.play();
        }
    }
    if (Input.isJustDown(Key.SPACE_BAR))
    {
        if (music.isPaused())
        {
            music.resume();
        }
        else
        {
            music.pause();
        }
    }
    if (Input.isJustDown(Key.Q)) notes[0].play(volume, balance);
    if (Input.isJustDown(Key.W)) notes[1].play(volume, balance);
    if (Input.isJustDown(Key.E)) notes[2].play(volume, balance);
    if (Input.isJustDown(Key.R)) notes[3].play(volume, balance);
    if (Input.isJustDown(Key.T)) notes[4].play(volume, balance);
    if (Input.isJustDown(Key.Y)) notes[5].play(volume, balance);
    if (Input.isJustDown(Key.U)) notes[6].play(volume, balance);
    if (Input.isJustDown(Key.I)) notes[7].play(volume, balance);
}

function render() {
    // Clear to black
    Renderer.clear(Color.BLACK);

    SpriteBatch.begin();

    SpriteBatch.drawText(font, "Press ^9901^999 to play sound1 at 100% speed", new Vector2(10, 10));
    SpriteBatch.drawText(font, "Press ^9902^999 to play sound1 at 50% speed", new Vector2(10, 30));
    SpriteBatch.drawText(font, "Press ^9903^999 to play sound1 at 200% speed", new Vector2(10, 50));
    SpriteBatch.drawText(font, "Press ^9904^999 to play a random sound", new Vector2(10, 70));
    SpriteBatch.drawText(font, "Press ^9905^999 to play sound5", new Vector2(10, 90));
    if (loopingSound.isPlaying())
    {
        SpriteBatch.drawText(font, "Press ^9906^999 to start/stop looping sound (^099Playing^999)", new Vector2(10, 110));
    }
    else
    {
        SpriteBatch.drawText(font, "Press ^9906^999 to start/stop looping sound", new Vector2(10, 110));
    }
    SpriteBatch.drawText(font, "Press ^9907^999 to play cue file", new Vector2(10, 130));
    SpriteBatch.drawText(font, "Press ^990qwertyui^999 to do music", new Vector2(10, 150));
    SpriteBatch.drawText(font, "Press ^9909^999 to play/stop music", new Vector2(10, 170));

    var screenH = Renderer.getResolution().y;
    SpriteBatch.drawText(font, "Hold ^990Left Arrow^999 to on left channel", new Vector2(10, screenH - 50));
    SpriteBatch.drawText(font, "Hold ^990Right Arrow^999 to on right channel", new Vector2(10, screenH - 30));

    SpriteBatch.end();
}
